data/slideseq2_olfactory_bulb/slideseq2_olfactory_bulb.py

```diff
@@ -0,0 +1,166 @@
+    "https://ftp.cngb.org/pub/SciRAID/stomics/STDS0000172/stomics/STSP0005386/GSM5173938_OB1_Slide14_processed.h5ad"
+    "https://ftp.cngb.org/pub/SciRAID/stomics/STDS0000172/stomics/STSP0005387/GSM5173939_OB1_Slide15_processed.h5ad"
+    "https://ftp.cngb.org/pub/SciRAID/stomics/STDS0000172/stomics/STSP0005388/GSM5173940_OB1_Slide16_processed.h5ad"
+    "https://ftp.cngb.org/pub/SciRAID/stomics/STDS0000172/stomics/STSP0005389/GSM5173941_OB1_Slide17_processed.h5ad"
+    "https://ftp.cngb.org/pub/SciRAID/stomics/STDS0000172/stomics/STSP0005390/GSM5173942_OB1_Slide18_processed.h5ad"
+    "https://ftp.cngb.org/pub/SciRAID/stomics/STDS0000172/stomics/STSP0005391/GSM5173943_OB1_Slide19_processed.h5ad"
+    "https://ftp.cngb.org/pub/SciRAID/stomics/STDS0000172/stomics/STSP0005392/GSM5173944_OB1_Slide20_processed.h5ad"
+]
+
+
+META_DICT = {"technology": "Slideseq2"}
+LICENSE = """ """
+SAMPLE_COLUMNS = [
+    "patient",
+    "sample",
+    "position",
+    "replicate",
+    "n_clusters",
+    "directory",
+]
+
+
+def download_links(links, temp_dir):
+    for link in links:
+        try:
+            response = urllib.request.urlopen(link)
+            # Extract filename from the URL
+            filename = os.path.join(temp_dir, urlparse(link).path.split("/")[-1])
+            with open(filename, "wb") as file:
+                file.write(response.read())
+            print(f"Downloaded: {filename}")
+        except Exception as e:
+            print(f"Error downloading {link}: {e}")
+
+
+def process_adata(adata_path, output_folder, iteration, samples_list):
+    adata = anndata.read_h5ad(adata_path)
+
+    # extract sample name from path
+    sample = os.path.basename(adata_path).split("_")[0]
+    complete_path = os.path.join(output_folder, sample)
+    print(f"Writing {sample}")
+    os.makedirs(complete_path, exist_ok=True)
+
+    # Observations
+    obs = adata.obs.copy()
+    obs["selected"] = "true"
+    obs.loc[obs.layer == "unknown", "selected"] = "false"
+    obs.to_csv(f"{complete_path}/observations.tsv", sep="\t", index_label="")
+
+    # Features
+    vars = adata.var.copy()
+    vars["selected"] = "true"
+    vars.to_csv(f"{complete_path}/features.tsv", sep="\t", index_label="")
+
+    # Coordinates
+    if "spatial" in adata.obsm:
+        coords = pd.DataFrame(adata.obsm["spatial"], columns=["x", "y"])
+    coords.index = adata.obs.index
+    coords.to_csv(f"{complete_path}/coordinates.tsv", sep="\t", index_label="")
+
+    # Matrix
+    scipy.io.mmwrite(f"{complete_path}/counts.mtx", adata.X)
+
+    # Write labels.tsv
+    labels = adata.obs["layer"]
+    labels.to_csv(f"{complete_path}/labels.tsv", sep="\t", index_label="")
+
+    # add info for sample.tsv
+    sample_data_basis = {
+        "patient": "0",
+        "sample": "0",
+        "position": iteration,
+        "replicate": "0",
+        "n_clusters": adata.obs.layer.nunique() - 1,  # -1 for unknown which we exclude
```

Comment:
Is this the convention we used? Maybe we should use NaN. Not sure how this was handled in other datasets

Reply:
Good point, will double check this

Reply:
Coming back to this, I think this was an missunderstanding. What is happening is that the authors annotated certain cells as unkown instead of filtering.
In keeping with the conventions we set for Spacehack I am also keeping the cells but setting them as selected == "false".
In line 105 I am merely calculating the number of clusters without the garbage cluster